File: Python/DesignPatterns/ObjectPool.py
# Object pool demo in Python

# The base vector class:
class Vector3D(object):
    def __init__(self):
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0

    def Set(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z
    
    def Clean(self):
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0


# The Laser Beam class:
class LaserBeam(object):
    def __init__(self):
        self.position = Vector3D()
    
    def Clean(self):
        self.position.Clean()


# The Laser Beam Object Pool:

class LaserBeamPool(object):
    def __init__(self, reserve = 5, maxCapacity = 100):
        self.activeList = []
        self.reserveList = []

        self.maxCapacity = maxCapacity
        self.numberActive = 0
        self.numberReserved = 0
        self._InitializeReserve(reserve)
        
    def _InitializeReserve(self, reserve):
        for i in range(reserve + 1):
            tmp = LaserBeam()
            self.reserveList.append(tmp)
            self.numberReserved += 1
    
    def CanSpawn(self):
        return self.numberActive < self.maxCapacity

    def GetLaserBeam(self):
        '''
            NOTICE:
            In your code, make sure to check the spawn limit.
            you can either:
                1. Throw an exception
                2. Return None
                3. Spawn a new zombie anyways.

                We'll return None.
        '''
        if self.numberActive >= self.maxCapacity:
            return None
        
        if self.numberReserved == 0:
            tmp = LaserBeam()
            self.reserveList.append(tmp)
            self.numberReserved += 1
        
        laser = self.reserveList.pop()
        self.numberReserved -= 1

        self.activeList.append(laster)
        self.numberActive += 1

        laser.Clean()

        return laser

    def ReturnLaserBeam(self, laser):
        if laser in self.activeList:
            self.activeList.remove(laser)
            self.numberActive -= 1
            
            self.reserveList.append(laser)
            self.numberReserved += 1

File: CSharp/DesignPatterns/ObjectPool.cs
// Object Pool Demo in C#
using System.Collections.Generic;
using System;

// The base Zombie class:
public class Zombie
{
    private const float START_HEALTH = 100.0f;

    private float health;

    // ... More variables for the zombie:

    public Zombie()
    {
        this.Clean();
    }

    // This seems only slightly ironic:
    public void IsAlive()
    {
        return health > 0.0f;
    }

    public void TakeDamage(float damage)
    {
        health -= damage;
    }

    // This resets the zombie to a "factory new" setting:
    public void Clean()
    {
        this.health = START_HEALTH;
    }

}

public class ZombiePool
{
    protected List<Zombie> activeList;
    protected List<Zombie> reserveList;

    int numberActive;
    int numberReserved;
    int spawnCap;

    public ZombiePool(int reserve = 5, int spawnCap = 20)
    {
        this.activeList = new List<Zombie>();
        this.reserveList = new List<Zombie>();

        this.numberActive = 0;
        this.numberReserved = 0;
        this.spawnCap = spawnCap;

        InitializeReserve(reserve);
    }

    private void InitializeReserve(int reserve)
    {
        for(int i = 0; i < reserve; i++)
        {
            Zombie zomb = new Zombie();
            this.reserveList.Add(zomb);
            this.numberReserved++;
        }
    }

    public bool CanSpawn()
    {
        return this.numberActive < this.spawnCap;
    }

    public Zombie GetZombie()
    {
        /*
            NOTICE:
            In your code, make sure to check the spawn limit.
            you can either:
                1. Throw an exception
                2. Return null
                3. Spawn a new zombie anyways.

                We'll return null.
        */

        if(this.numberActive >= spawnCap)
        {
            return null;
        }

        if(this.numberReserved == 0)
        {
            Zombie tmp = new Zombie();
            this.reserveList.Add(tmp);
            this.numberReserved++;
        }

        Zombie zomb = this.reserveList[0];
        this.reserveList.RemoveAt(0);
        this.numberReserved--;

        this.activeList.Insert(0, zomb);
        this.numberActive++;

        zomb.Clean();

        return zomb;

    }

    public void ReturnZombie(Zombie zomb)
    {
        if(this.activeList.Remove(zomb))
        {
            this.numberActive--;

            this.reserveList.Insert(0, zomb);
            this.numberReserved++;
        }
    }

}